### PR TITLE
Lint for redundant positional arguments.

### LINKED ISSUE #1768: Lint for pointless/bad positional arguments.
In cases like https://github.com/Citadel-Station-13/Citadel-Station-13-RP/blob/69ce25bd79b41b7800e33c45ccb9a094a280c5ff/code/modules/gamemaster/actions/atmos_leak.dm#L20-L23
```dm
...
severity = pickweight(EVENT_LEVEL_MUNDANE = 8,
EVENT_LEVEL_MODERATE = 5,
EVENT_LEVEL_MAJOR = 3
)
...
// Event defines.
#define EVENT_LEVEL_MUNDANE        1
#define EVENT_LEVEL_MODERATE    2
#define EVENT_LEVEL_MAJOR        3
...
```

After preprocessing, this is
```dm
// This runtimes if the argument order is different.
severity = pickweight(1 = 8,
2 = 5,
3 = 3
)
```
which is equivalent to
```dm
severity = pickweight(8, 5, 3)
```
this is almost certainly unintentional, undoubtably does not do what the programmer intended (a BYOND classic frankly), and should probably be linted for.

### PR DESCRIPTION
**Pull request by moonheart08**
Fixes #1768, adds a new lint.

```
Warning OD2210 at code/modules/gamemaster/actions/atmos_leak.dm:20:45: The argument at index 1 is a positional argument with a redundant index (`1 = value` at argument 1). This does not function like a named argument and is likely a mistake.
Warning OD2210 at code/modules/gamemaster/actions/atmos_leak.dm:21:24: The argument at index 2 is a positional argument with a redundant index (`2 = value` at argument 2). This does not function like a named argument and is likely a mistake.
Warning OD2210 at code/modules/gamemaster/actions/atmos_leak.dm:22:21: The argument at index 3 is a positional argument with a redundant index (`3 = value` at argument 3). This does not function like a named argument and is likely a mistake.
```

### CODE REVIEW

**Comment by ike709** [comment]
Unit tests plz

File: DMCompiler/DM/DMExpression.cs [hunk 1]
```diff
@@ -129,6 +136,7 @@ sealed class ArgumentList {
                 _isKeyed = true;
 
             Expressions[argIndex] = (name, value);
+
```
Comment (wixoaGit) [inline DMCompiler/DM/DMExpression.cs:139]: Unnecessary empty line
```suggestion
```

File: DMCompiler/DM/DMExpression.cs [hunk 2]
```diff
@@ -195,7 +203,8 @@ sealed class ArgumentList {
 
         DMComplexValueType paramType = param.ExplicitValueType ?? DMValueType.Anything;
 
-        if (!expr.ValType.IsAnything && !paramType.MatchesType(expr.ValType)) {
+
+        if ((!expr.ValType.IsAnything && !paramType.MatchesType(expr.ValType))) {
```
Comment (wixoaGit) [inline DMCompiler/DM/DMExpression.cs:207]: Unnecessary empty line & parenthesis
```suggestion
        if (!expr.ValType.IsAnything && !paramType.MatchesType(expr.ValType)) {
```

File: DMCompiler/DM/DMExpression.cs [hunk 3]
```diff
@@ -109,7 +109,14 @@ sealed class ArgumentList {
                     break;
                 case Expressions.Number keyNum:
                     //Replaces an ordered argument
-                    argIndex = (int)keyNum.Value;
+                    var newIdx = (int)keyNum.Value - 1;
+
+                    if (newIdx == argIndex) {
+                        DMCompiler.Emit(WarningCode.PointlessPositionalArgument, key.Location,
+                            $"The argument at index {argIndex + 1} is a positional argument with a redundant index (`{argIndex + 1} = value` at argument {argIndex + 1}). This does not function like a named argument and is likely a mistake.");
```
Comment (wixoaGit) [inline DMCompiler/DM/DMExpression.cs:116]: We avoid using backticks in compiler messages due to it causing formatting issues when pasted elsewhere
```suggestion
                            $"The argument at index {argIndex + 1} is a positional argument with a redundant index (\"{argIndex + 1} = value\" at argument {argIndex + 1}). This does not function like a named argument and is likely a mistake.");
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]
I'd also like for there to be unit tests. You can make one by `#pragma`-ing the lint to an error and purposefully triggering it.

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**